// @zxing/library/esm5/core/Exception.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Exception {
    constructor(type, message) {
        this.type = type;
        this.message = message;
    }
    getType() {
        return this.type;
    }
    getMessage() {
        return this.message;
    }
    static isOfType(ex, type) {
        return ex.type === type;
    }
}
Exception.IllegalArgumentException = 'IllegalArgumentException';
Exception.NotFoundException = 'NotFoundException';
Exception.ArithmeticException = 'ArithmeticException';
Exception.FormatException = 'FormatException';
Exception.ChecksumException = 'ChecksumException';
Exception.WriterException = 'WriterException';
Exception.IllegalStateException = 'IllegalStateException';
Exception.UnsupportedOperationException = 'UnsupportedOperationException';
Exception.ReedSolomonException = 'ReedSolomonException';
Exception.ArgumentException = 'ArgumentException';
Exception.ReaderException = 'ReaderException';
exports.default = Exception;
